Start sharing implementation

coverbeck opened the pull request:
This is a WIP; looking for some feedback and discussion. Some random thoughts and notes:

* There is the case of a FireCloud user calling Dockstore with a Google token, where the user hasn't registered with Dockstore. This currently won't work, as `SimpleAuthenticator:authenticate()` always looks up an existing user.
* Need an endpoint to indicate if a user could share a workflow.  I considered updating the User metadata object to have this, but that seems to all come from a database, and the second item below needs to be determined dynamically:
   1. Must be signed in with or have account linked to Google.
   2. Must have the account registered with SAM (we could do this for the user, but I don't think we should).
    3. Must be a hosted workflow.
* Not sure if some of this could have implemented with annotations.
* It may not be necessary to have both Roles and Actions.

┆Issue Number: DOCK-17

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/InMemoryPermissionsImpl.java:
```diff
@@ -0,0 +1,157 @@
+        List<Permission> list = new ArrayList<>();
+        if (entry instanceof Workflow) {
+            Workflow workflow = (Workflow)entry;
+            if (workflow.getMode() == WorkflowMode.HOSTED) {
```
Its not clear to me, why are hosted workflows treated specially here? 
My guess: we were planning on testing out permissions with hosted workflows first. But it makes more sense to me to disable the GUI for permissions on anything but hosted workflows rather than treat it as a special case in the webservice. Could be misinterpreting. 

If it is necessary to write temp code that will be removed later, add a `//DODO` comment explaining

coverbeck replied:
I will remove this section. This class was lagging; I had started with it and was playing with it before I got SAM running locally, then when I got SAM working I stopped working on this.

It needs to be correct for testing, so I will go through it again.

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/sam/SamPermissionsImpl.java:
```diff
@@ -0,0 +1,257 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(SamPermissionsImpl.class);
+
+    private static Map<String, Role> samPermissionMap = new HashMap<>();
```
Document maps when they use types like Strings, Pairs, etc. where the value isn't obvious

coverbeck replied:
Fixing in next commit

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:
```diff
@@ -691,10 +721,70 @@ public Workflow getWorkflowByPath(@ApiParam(hidden = true) @Auth User user, @Api
+        return this.permissionsInterface.getPermissionsForWorkflow(user, workflow);
+    }
+
+    @PATCH
```
Could this just be a Put since it seems idempotent

coverbeck replied:
I made it PATCH because it's only modifying a part of the collection. For example, if the permissions on a resource are:
```
[
  {
    "email": "jdoe@ucsc.edu",
    "role": "OWNER"
  },
  {
    "email": "jane.doe@gmail.com",
    "role": "READER"
  }
]
```
Then I could add another user without having to specify the whole body (and that would less likely for users to clobber each other's changes).

coverbeck commented on swagger-java-sam-client/pom.xml:
```diff
@@ -0,0 +1,306 @@
+                            <goal>generate</goal>
+                        </goals>
+                        <configuration>
+                            <inputSpec>https://sam.dsde-prod.broadinstitute.org/api-docs.yaml</inputSpec>
```
> Is there a current use case for a FireCloud user interacting with Dockstore without ever having logged in?

My understanding is that Dockstore will replace the FireCloud methods repository, so that in the FireCloud UI they will be eventually fetching and storing all workflows/methods in Dockstore. I thought we needed to support this from the start, but we should clarify with Broad.

coverbeck replied:
> This is always reading off the live server, raising the spectre of our builds suddenly failing if their server is upgraded or offline. Is this checked in and tagged anywhere in their repos?

It is checked in, but unfortunately they don't tag their releases. I'm worried that we could end up generating a client for something that hasn't yet been deployed. We could broach this at the next meeting with FireCloud.

denis-yuen replied:
Probably worthwhile to bring it up, being able to say which version of sam we tested against is import. In the meantime, you can point at checked in content by commit
https://help.github.com/articles/getting-permanent-links-to-files/#press-y-to-permalink-to-a-file-in-a-specific-commit

coverbeck replied:
OK, changing to point to a GitHub commit in next push.

coverbeck commented on dockstore-webservice/src/test/java/io/dockstore/webservice/SimpleAuthenticatorTest.java:
```diff
@@ -0,0 +1,16 @@
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
Fixed.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/InMemoryPermissionsImpl.java:
```diff
@@ -0,0 +1,157 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */package io.dockstore.webservice.permissions;
```
Will be fixed in next push.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/InMemoryPermissionsImpl.java:
```diff
@@ -0,0 +1,157 @@
+ */
+public class InMemoryPermissionsImpl implements PermissionsInterface {
+
+    private final Map<String, Map<String, Role>> map = new ConcurrentHashMap<>();
```
Comment added in next push, and gave it a better name, `resourceToUsersAndRolesMap`.

coverbeck commented on dockstore-webservice/src/main/resources/swagger.yaml:
```diff
@@ -3400,6 +3400,63 @@ paths:
+        required: true
+        type: "string"
+        enum:
+        - "WRITE"
```
I tend to agree, and had my doubts about this in my PR comments. This cements it... I will work on an implementation without actions.

coverbeck replied:
I'm now having second thoughts about agreeing. :)

The nice thing about actions is that the UI can query whether a user can perform a certain action and enable/disable UI elements accordingly:

```
GET https://dockstore.org:8443/workflows/path/workflow/{repository}/actions?action=write
```

vs if the UI queried for a role:

```
GET https://dockstore.org:8443/workflows/path/workflow/{repository}/role
```

Then the UI would have to know what actions a role could perform.

Maybe a solution is to get rid of roles and keep actions? Although I don't really like that from the perspective of managing actions for a resource.

Maybe best to mull on it a little bit more.

coverbeck replied:
Thought about this some more, and I think I'll get rid of the actions endpoint. Which means roles alone should work. 

For the UI, we can add OPTIONS endpoints that will let the UI query whether the user has permissions to do something.

denis-yuen replied:
One thought, I'm opposed to having the two as they are, as two discrete clases/enums. 

However, if the connection were made more explicit, like the Role enum declaring the action enum as an inner enum and storing it (a bit like https://github.com/ga4gh/dockstore/blob/develop/dockstore-common/src/main/java/io/dockstore/common/Registry.java stores friendly names) then that might be slightly better. That could aid in the options endpoints

cricketsloan commented:
➤ Cricket Sloan commented:

this is a pull request, unito likely has an issue

[~casloan]

coverbeck commented:
I have made several updates to this and removed the WIP, so that, pending review, it can get merged.

Here are the major changes I made since the initial commit:

* Moved Action to be an enum inside Role.java
* Added integration tests
* Got rid of a new endpoint to query permissions; instead now using OPTIONS which returns the HTTP methods that are allowed. I only added the operation to a couple of endpoints; the idea is that if you make an unauthorized OPTIONS call, all possible methods are returned, whereas if you make an authorized OPTIONS call, only the methods applicable to that user are returned. The problem I ran into is that the browser is sending unauthenticated OPTIONS calls as part of the CORS negotiation, so I need to return all possible methods if unauthenticated.

By default, this implementation will use NoOpPermissionsImpl.java, which should leave current behavior unchanged. dockstore.yml needs to be updated with `authorizerType` to use either the In Memory or SAM implementation.

Todo, which I will tackle in a future PR(s):

* Allowing Google users who haven't registered with Dockstore to access shared workflows.
* Perhaps endpoint(s) to indicate if sharing is enabled, and if not, why not.

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/PermissionsInterface.java:
```diff
@@ -0,0 +1,102 @@
+ * <p>
+ *     If we extend sharing to tools, should we create additional methods for tools here,
+ *     e.g., <code>setPermission(Tool tool, User requester, Permission permission)</code>,
+ *     or should we instead make this interface more generic, and just require a
```
(change to comment) Probably just make this generic in the future:
i.e. `List<Permission> setPermission(Entry entry, User requester, Permission permission);` instead of `List<Permission> setPermission(Workflow workflow, User requester, Permission permission);`

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractHostedEntryResource.java:
```diff
@@ -138,6 +152,58 @@ public T editHosted(@ApiParam(hidden = true) @Auth User user,
+     * @param entryId
+     * @return
+     */
+    @OPTIONS
```
Neat

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AuthenticatedResourceInterface.java:
```diff
@@ -89,4 +89,52 @@ default void checkUser(User user, long id) {
+    /**
+     * Checks if a user can read an entry. Default implementation
+     * is to invoke <code>checkUser</code>. Implmenations that support
+     * more nuanched sharing should override.
```
nuanced

denis-yuen commented on swagger-java-client/src/main/resources/swagger-templates/ApiClient.mustache:
```diff
@@ -0,0 +1,820 @@
+        response = invocationBuilder.method("PATCH", entity);
+      } else if ("HEAD".equals(method)) {
+        response = invocationBuilder.head();
+      } else if ("OPTIONS".equals(method)) {
```
This template is a bit big, could you add comments to indicate the change that was made?
This would aid updates in the future that might need to merge these changes with an update to swagger codegen.
(Hoping it was just this couple of lines)

Alternatively, comment pointing to a PR to swagger codegen?

coverbeck replied:
I first tried to go down the PR route, but the instructions in the repo on contributing vs the state of tests didn't seem to match. Plus the master branch is on 3.0. It seemed like it would be a bit of work, so I decided to go with just doing our own template.

Good idea on the comments; I will add comments to show what I changed (it's just 2 lines).

agduncan94 commented on dockstore-webservice/src/main/java/io/dockstore/webservice/SimpleAuthenticator.java:
```diff
@@ -46,10 +49,35 @@ public SimpleAuthenticator(TokenDAO dao, UserDAO userDAO) {
+     * @return
+     */
+    static boolean isGoogleToken(String credentials) {
+        return credentials != null && credentials.startsWith("ya29.");
```
Could use a comment about the prefix ya29.

coverbeck replied:
Good comment, which opened a can of worms. :) I decided to get rid of this method entirely, because it was a fragile check. I changed the `authenticate` method to instead assume it's a Dockstore token and look it up; if that fails, then we try to see if it's a Google token.

coverbeck commented on dockstore-webservice/src/test/java/io/dockstore/webservice/SimpleAuthenticatorTest.java:
```diff
@@ -0,0 +1,17 @@
+        assertFalse(SimpleAuthenticator.isGoogleToken(null));
+        assertFalse(SimpleAuthenticator.isGoogleToken(""));
+        assertFalse(SimpleAuthenticator.isGoogleToken("6b9f637671356a3919dbed3d1b007a0af257d2e9d80f868f9954f2ff63n8858c"));
+        assertTrue(SimpleAuthenticator.isGoogleToken("ya29.GlzNBW56KugKqe38nCQGZjh_2Y_VGvqEjjq1U0wQ1zcqqFXYJtRb-rekGgOSrR7FlHenn9ULZGSHfBoJYouKM4iI7uSWryIpobWxE6JVLEIyL-ZRBlBbqK961QPq8A"));
```
Yes, for the Google token, I took an expired Google token and changed some characters; for the other one, I also changed some characters.
